acp/identity/identity.go

```diff
@@ -8,34 +8,48 @@
 // by the Apache License, Version 2.0, included in the file
 // licenses/APL.txt.
 
-/*
-Package identity provides defradb identity.
-*/
-
 package identity
 
-import "github.com/sourcenetwork/immutable"
+import (
+	cosmosSecp256k1 "github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	"github.com/cosmos/cosmos-sdk/types"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/sourcenetwork/immutable"
+)
 
-// Identity is the unique identifier for an actor.
-type Identity string
+// None specifies an anonymous actor.
+var None = immutable.None[Identity]()
 
-var (
-	// None is an empty identity.
-	None = immutable.None[Identity]()
-)
+type Identity struct {
+	// PublicKey is the identity public key.
+	PublicKey *secp256k1.PublicKey
+	// PrivateKey is the identity private key.
+	PrivateKey *secp256k1.PrivateKey
+	// Address is the identity address.
+	Address string
+}
+
+// FromPrivateKey returns a new identity using the given private key.
+func FromPrivateKey(privateKey *secp256k1.PrivateKey) immutable.Option[Identity] {
+	pubKey := privateKey.PubKey()
+	return immutable.Some(Identity{
+		Address:    AddressFromPublicKey(pubKey),
+		PublicKey:  pubKey,
+		PrivateKey: privateKey,
+	})
+}
 
-// New makes a new identity if the input is not empty otherwise, returns None.
-func New(identity string) immutable.Option[Identity] {
-	// TODO-ACP: There will be more validation once sourcehub gets some utilities.
-	// Then a validation function would do the validation, will likely do outside this function.
-	// https://github.com/sourcenetwork/defradb/issues/2358
-	if identity == "" {
-		return None
-	}
-	return immutable.Some(Identity(identity))
+// FromPublicKey returns a new identity using the given public key.
+func FromPublicKey(publicKey *secp256k1.PublicKey) immutable.Option[Identity] {
+	return immutable.Some(Identity{
+		Address:   AddressFromPublicKey(publicKey),
+		PublicKey: publicKey,
+	})
 }
 
-// String returns the string representation of the identity.
-func (i Identity) String() string {
-	return string(i)
+// AddressFromPublicKey returns the unique address of the given public key.
+func AddressFromPublicKey(publicKey *secp256k1.PublicKey) string {
+	pub := cosmosSecp256k1.PubKey{Key: publicKey.SerializeCompressed()}
+	// conversion from well known types should never cause a panic
+	return types.MustBech32ifyAddressBytes("cosmos", pub.Address().Bytes())
```

Comment:
thought:  This looks like it might actually panic if `"cosmos"` is no longer the prefix (it is actually being changed at the moment, so that's not without precedent).

todo: There is a const for this that you can use instead of a magic string, please use `"github.com/sourcenetwork/sourcehub/app"`.`AccountAddressPrefix`

Reply:
Using that causes a panic `panic: invalid Bech32 prefix; expected cosmos, got source`

Reply:
Ah okay, using the const might have to wait on Shahzad's PR then, and I've been chatting with Bruno today - he might change the source hub sdk so that Defra controls the prefix (in which case we can either own a const, or have a config option for this).